### ISSUE TITLE
Proposal: Relax the strong and other inline rules

### ISSUE THREAD

**Issue by criloz** [jgm/djot#319]
## Current behavior

The current strong rule `*..*` only works if there is no white space after the first `*` and before the last `*`, there are other rules similar to this, also it requires both, one starting `*` and one ending `*` to be effective

- ` *sds *` not strong
-  ` *sds ` not strong

## Problem

this creates a weird typing behavior in djot editors (I am currently working in one) 
so the user do this typing sequence 

1. `*hello*`
2.  `*hello *`
3.  `*hello world*`
4.  `*hello world *`
5.  `*hello world ...*`

As you can see, every time that the user type a space the whole span get marked as not strong until the user start typing other word and the cycle start again, giving a weird **intermittent** and not **continuous** user experience

## Proposals

1. allow whitespace before the last `*`, do the same for other similar rules

Why not whitespace after the first `*`?  Because it will create an ambiguity with bullet list.
Another option is to allow whitespace after the first `*`, but in case of ambiguity with bullet list give priority to the bullet list.

2.  [optionally] just considered anything that follows an asterisk without space a strong structure, that is close when `*` match, this repeated in other similar inline structures. 
   
You can see this behavior in the obsidian editor that use Markdown, something like `**sdasd` will be marked as strong even if there is not a final matching `**`. 
   
The second proposal is marked as optional because there is no need to be in the spec for a theoretical djot editor to have this behavior, and they can post-process the file and produce a valid djot output anyway.

**Comment by Omikhleia** [comment]
> Allow whitespace before the last `*`,

But then, regarding: `Triconsonantal roots *ZBD or *ZBN are different...`

Is it showing any strong content, is is this just two star-prefixed words (as often seen e.g. for reconstructions in linguistics)?

**Comment by vassudanagunta** [comment]
> this creates a weird typing behavior in djot editors

This is an editor UX problem, not a syntax design problem.

**Comment by criloz** [comment]
@vassudanagunta not always, some editors use the parser to decorate subregion of the text and assign some classes, like codemirror, basically if you want to build an editor with djot this rule will by default create this intermittent behavior, the only option will be implementing a new djot parser with those relaxed rules and do the fixes in a post-processing step, but it means that the official implementation or the official grammar could not be used by default. not a problem but is something that could be considered

**Comment by vassudanagunta** [comment]
@criloz, respectfully, that doesn't change my point. Even more so for a plain text syntax. Whatever the desired editor UX, it should and _can_ be solved without changing the syntax.

**Comment by criloz** [comment]
@Omikhleia your example will render like 

```html
Triconsonantal roots <strong>ZBD or </strong> ZBN are different...
```

**Comment by criloz** [comment]
@vassudanagunta I understand your concerns, those are just suggestion, they are open discussion and I think should not  be discarded so quickly, it allows anyone to just use the default djot parser without need to write a new one to implement an editor that does not feel hacky, it can potentially increase the speed of adaptability of djot, also by my experience, more relaxed rules create also simple parsers.

**Comment by vassudanagunta** [comment]
> should not be discarded so quickly

I apologize that I gave you that impression. I should have pointed out that there are important reasons for those rules, and those rules cannot be discarded so quickly, especially for editor implementation convenience. 

I can't state all the reasons off the top of my head, but one of the reasons has to do with ambiguity when inline spans are potentially nested.

You should be able to find all the reasons and nuances behind the rules in the CommonMark [spec](https://spec.commonmark.org/0.31.2/#emphasis-and-strong-emphasis) and by [searching](https://talk.commonmark.org/search?context=topic&context_id=3478&q=emphasis%20&skip_context=true) the forum, from which djot inline rules descend (i believe). 

As an example of how complicated it can be, see [Emphasis with CJK punctuation #650](https://github.com/commonmark/commonmark-spec/issues/650).

In other words, it's already impossible to design the syntax to support all use case while remaining maximally intuitive, coherent and elegant. Adding graphical editor UX concerns to the requirement will have bad consequences.

**Comment by criloz** [comment]
@vassudanagunta thank for the context, and you are completely right, it will be impossible to do it, rethinking again in my particular case a possible solution would be intercepting the intention when the user inserts a `*` that is not Preceded by `\` and actually insert `{**}`, so this behavior is not present also replace all the `*strong*` in a preprocessor step by `{*strong*}`

**Comment by Omikhleia** [comment]
> @Omikhleia your example will render like
> 
> ```
> Triconsonantal roots <strong>ZBD or </strong> ZBN are different...
> ```

Ah. So would I have to backslash-escape all such starred-roots in a document? This doesn't feel intuitive, esp. for a lightweight markup language. More relaxed rules often cause ambiguity, and contraints for the typist.

**Comment by bpj** [comment]
@Omikhleia I currently have to backslash escape such "linguistic" asterisks in Pandoc Markdown and it is  a PITA as I keep forgetting. Due to a quirk in Pandoc's Markdown parser `_*this_` works, but it breaks again when you change it into `_*this, *that_` so I have to train myself to remember to escape :-( I much prefer the (current) djot rules!

**Comment by bpj** [comment]
@vassudanagunta 

> , it should and can be solved without changing the syntax.

“the syntax of the markup language” to be clear.

Yes the syntax of the language should absolutely not be changed because some editor gets a bad UX because it is lazy and piggybacks its highlighting parsing on the language parser. There is a good reason to keep highlighting parsers separate.
